[Eval] Add coding progress assessment eval

danielstrizhevsky opened the pull request:
# Thank you for contributing an eval! ♥️

🚨 Please make sure your PR follows these guidelines, **failure to follow the guidelines below will result in the PR being closed automatically**. Note that even if the criteria are met, that does not guarantee the PR will be merged nor GPT-4 access be granted. 🚨

**PLEASE READ THIS**:

In order for a PR to be merged, it must fail on GPT-4. We are aware that right now, users do not have access, so you will not be able to tell if the eval fails or not. Please run your eval with GPT-3.5-Turbo, but keep in mind as we run the eval, if GPT-4 gets higher than 90% on the eval, we will likely reject it since GPT-4 is already capable of completing the task.

We plan to roll out a way for users submitting evals to see the eval performance on GPT-4 soon. Stay tuned! Until then, you will not be able to see the eval performance on GPT-4. **Starting April 10, the minimum eval count is 15 samples, we hope this makes it easier to create and contribute evals.**

Also, please note that we're using **Git LFS** for storing the JSON files, so please make sure that you move the JSON file to Git LFS before submitting a PR. Details on how to use Git LFS are available [here](https://git-lfs.com).

## Eval details 📑

### Eval name

Code Progress

### Eval description

The model is given a coding interview question and a code block representing a snapshot of an interviewee's editor. The model is tasked with deciding whether the code block is

- Correct
- Incorrect
- Incomplete, and on the right track
- Incomplete, and on the wrong track

### What makes this a useful eval?

It's valuable for a model to understand whether its input is currently incomplete, or whether it is complete and ready for response. This is especially useful for real-time applications where a stream of data (e.g. someone's editor in a coding interview) is periodically sent to the model, with the expectation that the model decides when it's appropriate to respond.

## Criteria for a good eval ✅

Below are some of the criteria we look for in a good eval. In general, we are seeking cases where the model does not do a good job despite being capable of generating a good response (note that there are some things large language models cannot do, so those would not make good evals).

Your eval should be:

- [x] Thematically consistent: The eval should be thematically consistent. We'd like to see a number of prompts all demonstrating some particular failure mode. For example, we can create an eval on cases where the model fails to reason about the physical world.
- [x] Contains failures where a human can do the task, but either GPT-4 or GPT-3.5-Turbo could not.
- [x] Includes good signal around what is the right behavior. This means either a correct answer for `Basic` evals or the `Fact` Model-graded eval, or an exhaustive rubric for evaluating answers for the `Criteria` Model-graded eval.
- [x] **Include at least 15 high-quality examples.**

If there is anything else that makes your eval worth including, please document it below.

### Unique eval value

> Insert what makes your eval high quality that was not mentioned above. (Not required)

## Eval structure 🏗️

Your eval should

- [x] Check that your data is in `evals/registry/data/{name}`
- [x] Check that your YAML is registered at `evals/registry/evals/{name}.yaml`
- [x] Ensure you have the right to use the data you submit via this eval

(For now, we will only be approving evals that use one of the existing eval classes. You may still write custom eval classes for your own cases, and we may consider merging them in the future.)

## Final checklist 👀

### Submission agreement

By contributing to Evals, you are agreeing to make your evaluation logic and data under the same MIT license as this repository. You must have adequate rights to upload any data used in an Eval. OpenAI reserves the right to use this data in future service improvements to our product. Contributions to OpenAI Evals will be subject to our usual Usage Policies (<https://platform.openai.com/docs/usage-policies>).

- [x] I agree that my submission will be made available under an MIT license and complies with OpenAI's usage policies.

### Email address validation

If your submission is accepted, we will be granting GPT-4 access to a limited number of contributors. Access will be given to the email address associated with the commits on the merged pull request.

- [x] I acknowledge that GPT-4 access will only be granted, if applicable, to the email address used for my merged pull request.

### Limited availability acknowledgment

We know that you might be excited to contribute to OpenAI's mission, help improve our models, and gain access to GPT-4. However, due to the requirements mentioned above and the high volume of submissions, we will not be able to accept all submissions and thus not grant everyone who opens a PR GPT-4 access. We know this is disappointing, but we hope to set the right expectation before you open this PR.

- [x] I understand that opening a PR, even if it meets the requirements above, does not guarantee the PR will be merged nor GPT-4 access be granted.

### Submit eval

- [x] I have filled out all required fields of this form
- [x] I have used **Git LFS** for the Eval JSON data
- [ ] (Ignore if not submitting code) I have run `pip install pre-commit; pre-commit install` and have verified that `black`, `isort`, and `autoflake` are running when I commit and push

Failure to fill out all required fields will result in the PR being closed.

### Eval JSON data

Since we are using Git LFS, we are asking eval submitters to add in as many Eval Samples (at least 5) from their contribution here:

<details>
  <summary>View evals in JSON</summary>

  ### Eval
  ```jsonl
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = -1\n    for num in nums:```"}],"ideal":"[RIGHT TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = -1\n    for num in nums:\n        if num > la\n```"}],"ideal":"[RIGHT TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = -1\n    for num in nums:\n        if num > largest:\n            largest = n\n```"}],"ideal":"[RIGHT TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = -1\n    for num in nums:\n        if num > largest:\n            largest = num\n    return largest\n```"}],"ideal":"[CORRECT]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = -1\n    for num in nums:\n        if num < largest:\n            largest = num\n    return largest\n```"}],"ideal":"[INCORRECT]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = 100\n    for num in nums:\n        if num < largest:\n            largest = num\n    return largest\n```"}],"ideal":"[INCORRECT]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    largest = -1\n    for num in nums:\n        if num < largest:\n```"}],"ideal":"[WRONG TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of positive integers, return the largest integer. \nCode input:\n```\ndef get_max(nums):\n    estimate = nums[-1]**2```"}],"ideal":"[WRONG TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(nums):\n    duplicate = 0```"}],"ideal":"[RIGHT TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(nums):\n    duplicate = 0\n    for num in nums:\n        duplic```"}],"ideal":"[RIGHT TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(nums):\n    duplicate = 0\n    for num in nums:\n        duplicate ^= num\n```"}],"ideal":"[RIGHT TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(nums):\n    duplicate = 0\n    for num in nums:\n        duplicate ^= num\n    return duplicate\n```"}],"ideal":"[CORRECT]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(nums):\n    duplicate = 0\n    for num in nums:\n        duplicate &= num\n    return duplicate\n```"}],"ideal":"[INCORRECT]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(nums):\n    duplicate = 0\n    for num in nums:\n        duplicate &= num\n```"}],"ideal":"[WRONG TRACK]"}
{"input":[{"role":"system","content":"Act as a coding interviewer. You will receive a coding question and an input code block. The input code is a snapshot of the user's editor, so it may be incomplete. Decide whether the code is (1) correct, (2) incorrect, (3) in progress and on the right track, or (4) in progress but on the wrong track. Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK]."},{"role":"user","content":"Coding question: Given a list of integers where one integer appears exactly once and the others appear exactly twice, find the integer that appears once.\n\nCode input:\n```\ndef find_duplicate(s):\n    s = s[::-1]```"}],"ideal":"[WRONG TRACK]"}
```
</details>

usama-openai reviewed (approved):
Thanks for submitting this eval! This PR looks interesting. I'm approving this PR.

danielstrizhevsky commented:
Actually - realized these samples can have false negatives. If I'm looking for "includes `correct`" and the answer is `Incorrect.`, it will match the substring and count it as passing. I will try to fix that - let me know if there's an easy way you can think of.

danielstrizhevsky commented:
I've updated the prompt to instruct the model to respond the following way: `Your answer must be one of the following phrases, enclosed in square brackets: [CORRECT], [INCORRECT], [RIGHT TRACK], [WRONG TRACK].` I've updated the `ideal` responses to be these phrases with brackets.

This removes the false negatives (e.g. the model response is `incorrect` and the ideal response is `correct`, which would have passed because `correct` is a substring of `incorrect`). Now it tests for two things at the same time - correctly analyzing the progress/correctness of the code, as well as following directions on output formatting. Sometimes it responds with something like `[IN PROGRESS - RIGHT TRACK]`, which would fail an ideal response of `[RIGHT TRACK]`. I think this is fair.

usama-openai commented:
You should see GPT-4 API access enabled in your account in the next few days.
